File: packages/emitter/testcases/common/extensions/linq/ExtensionMethods.ts
import type { thisarg } from "@tsonic/core/lang.js";
import { int } from "@tsonic/core/types.js";
import { Console } from "@tsonic/dotnet/System.js";
import { List } from "@tsonic/dotnet/System.Collections.Generic.js";
import type { IEnumerable } from "@tsonic/dotnet/System.Collections.Generic.js";
import type { ExtensionMethods as Linq } from "@tsonic/dotnet/System.Linq.js";

type LinqSeq<T> = Linq<IEnumerable<T>>;

export function inc(x: thisarg<int>): int {
  return x + 1;
}

export function run(): void {
  const numbers = new List<int>();
  numbers.Add(1);
  numbers.Add(2);
  numbers.Add(3);
  numbers.Add(4);

  const xs = numbers as unknown as LinqSeq<int>;
  const doubled = xs
    .Where((n) => n % 2 === 0)
    .Select((n) => n * 2)
    .ToList();

  Console.WriteLine(doubled.Count);
  Console.WriteLine(inc(5));
}

File: packages/emitter/testcases/common/expected/extensions/linq/ExtensionMethods.cs
namespace TestCases.common.extensions.linq
{
    [global::Tsonic.Internal.ModuleContainerAttribute]
    public static class ExtensionMethods
    {
        public static int inc(this int x)
        {
            return x + 1;
        }

        public static void run()
        {
            var numbers = new global::System.Collections.Generic.List<int>();
            numbers.Add((int)1);
            numbers.Add((int)2);
            numbers.Add((int)3);
            numbers.Add((int)4);
            var xs = (global::System.Collections.Generic.IEnumerable<int>)numbers;
            var doubled = global::System.Linq.Enumerable.Select(global::System.Linq.Enumerable.Where(xs, (int n) => n % 2 == (int)0), (int n) => n * 2).ToList();
            global::System.Console.WriteLine(doubled.Count);
            global::System.Console.WriteLine(inc((int)5));
        }
    }
}
